packages/medusa/src/repositories/product.ts

```diff
@@ -224,40 +232,48 @@ export const ProductRepository = dataSource.getRepository(Product).extend({
       if (includeCategoryChildren) {
         const categoryRepository =
           this.manager.getTreeRepository(ProductCategory)
+
         const categories = await categoryRepository.find({
-          where: { id: In(categoryIds) },
+          where: {
+            id: In(categoryIds),
+            ...categoriesQuery,
+          },
         })
 
-        categoryIds = []
         for (const category of categories) {
           const categoryChildren = await categoryRepository.findDescendantsTree(
             category
           )
 
-          const getAllIdsRecursively = (productCategory: ProductCategory) => {
-            let result = [productCategory.id]
-
-            ;(productCategory.category_children || []).forEach((child) => {
-              result = result.concat(getAllIdsRecursively(child))
-            })
-
-            return result
-          }
-
           categoryIds = categoryIds.concat(
-            getAllIdsRecursively(categoryChildren)
+            fetchCategoryDescendantsIds(categoryChildren, categoriesQuery)
           )
         }
       }
 
-      joinMethod(
-        `${productAlias}.categories`,
-        "categories",
-        "categories.id IN (:...categoryIds)",
-        {
-          categoryIds,
+      if (categoryIds.length) {
+        const categoryAlias = "categories"
+        const joinScope = {
+          ...categoriesQuery,
+          id: categoryIds,
         }
-      )
+        const joinWhere = Object.entries(joinScope)
+          .map((entry) => {
+            if (Array.isArray(entry[1])) {
+              return `${categoryAlias}.${entry[0]} IN (:...${entry[0]})`
+            } else {
+              return `${categoryAlias}.${entry[0]} = :${entry[0]}`
+            }
```

Comment:
suggestion(non-blocking): Reduce to a single query
```suggestion
            const categoryInQuery = Array.isArray(entry[1]) ? entry[0] : [entry[0]]

            return `${categoryAlias}.${entry[0]} IN (:...${categoryInQuery})`
```

Reply:
@olivermrbl this one unforuntately didn't work. 

```
entry -  [ 'is_internal', false ]
entry -  [ 'is_active', true ]
entry -  [ 'id', [ 'category-with-product-id' ] ]
joinWhere -  categories.is_internal IN (:...is_internal) AND categories.is_active IN (:...is_active) AND categories.id IN (:...id)
joinScope -  {
  is_internal: false,
  is_active: true,
  id: [ 'category-with-product-id' ]
}
error:   value.map is not a function
TypeError: value.map is not a function
    at /Users/riqwanthamir/Projects/medusa/node_modules/typeorm/driver/postgres/PostgresDriver.js:605:22
    at String.replace (<anonymous>)
    at PostgresDriver.escapeQueryWithParameters (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/driver/postgres/PostgresDriver.js:598:19)
    at SelectQueryBuilder.getQueryAndParameters (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/query-builder/QueryBuilder.js:276:39)
    at SelectQueryBuilder.loadRawResults (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/query-builder/SelectQueryBuilder.js:2102:40)
    at SelectQueryBuilder.getRawMany (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/query-builder/SelectQueryBuilder.js:629:40)
    at SelectQueryBuilder.executeEntitiesAndRawResults (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/query-builder/SelectQueryBuilder.js:1948:18)
    at SelectQueryBuilder.getManyAndCount (/Users/riqwanthamir/Projects/medusa/node_modules/typeorm/query-builder/SelectQueryBuilder.js:833:47)
    at Object.<anonymous> (/Users/riqwanthamir/Projects/medusa/packages/medusa/dist/repositories/product.js:122:59)
    at step (/Users/riqwanthamir/Projects/medusa/packages/medusa/dist/repositories/product.js:44:23)
```

Reply:
Note, I didn't test my proposed changes but just added it to convey the idea. It might need some tweaking. That said, I am also fine keeping it out for now :)